sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Error related to the Simulator interface. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "EMU error: {}".  So including the words "error" or
+/// "Emu" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct EmuArgs {
+    param: String,
+    value: Option<String>,
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+pub enum EmuState {
+    Off,
+    On,
+    Busy,
+    Error,
+}
+
+/// A trait which represents a single GPIO pin.
+pub trait Emulator {
+    /// Aquire status of emulator
+    fn status(&self) -> Result<EmuState>;
+
+    /// Start emulator with provided arguments
```

Comment:
It is worth documenting here how start behaves if the current state is either On, Busy or Error.
Also, if you can specify content of two flash sections, but give only one to start(). Will the other be reset to "default", as the first call to start() after creation, or will the content of that flash section survive from the last time the emulator was "On"?  In case of the latter, is there a specified way of completely wiping the emulator status back to the same as it was when the session was created?  (Usually, the user could just stop the session and start a new, if they wanted that effect, but thinking of that emulator farm mentioned before, the user may not have shell access to the machine running the emulator, yet would want to wipe the state left from the previous user of the emulator instance.)
I do not think that `start()` should wipe flash sections not mentioned, as I could see use cases for upgrading the emulated code section of flash, while leaving the emulated filesystem section intact.  I also think that if `start()` leaves intact any section/setting not explicitly mentioned, that we need a separate `factory_reset()` method, or similar, which can be used to bring the emulator back to the same start as it was at session creation.

Reply:
I was thinking of some kind of resource_reset(...) like function.
When we have explicit resource/file reset then problem of base state occur, we need to store untouched copy of resource. Also some king of send_resource(...) function may be required to make it work.

IMHO: Sending file over network should be performed separately by real world prove solution like  scp or NFS... I would like to avoid building next network file system or FTP protocol only because we need to run simulation in Verilator.

Reply:
> IMHO: Sending file over network should be performed separately by real world prove solution like scp or NFS... I would like to avoid building next network file system or FTP protocol only because we need to run simulation in Verilator.
I fundamentally disagree.  I think it is highly useful to be able to equally do:
```
opentitantool --proxy localhost:9900 emulator start build/ti50.bin
```
and
```
opentitantool --proxy v1.lab.corp:9900 emulator start build/ti50.bin
```
With PR #11311, we will have that exact feature for flashing actual hardware.

I think we need to have a discussion with some people from the OpenTitan team to figure out the direction here.  I will try to set up a meeting, maybe Monday evening European time.

Reply:
Good idea, Monday is fine.

Reply:
I agree that this will be usefully, I have only problem on how we should achieve this.

Reply:
@cfrantz @jesultra
As agreed at today's meeting:
1) Enum extended with entry for file name
2) New flag for resource reset in fn start(..)